backend/owa.ts

```diff
@@ -33,26 +46,40 @@ export async function fetchSessionData(partition: string, url: string, interacti
       };
       let checkLoginFinished = async function() {
         try {
-          response = await session.fetch(url + 'sessiondata.ashx', { method: 'POST' });
+          response = await session.fetch(urlObj.toString() + 'sessiondata.ashx', { method: 'POST' });
           finish(await response.json());
         } catch (ex) {
         }
       }
-      let checkCanary = function(_event, cookie, _cause, removed) {
-        if (!removed && cookie.name == 'X-OWA-CANARY') {
-          checkLoginFinished();
+      let checkCanary = async function(_event, cookie, _cause, removed) {
+        // For Hotmail and personal accounts, check the path to the CANARY cookie.
+        if (cookie.domain == kHotmailServer &&
+          cookie.name == kCookieName &&
+          cookie.path?.match(reOWAPath)) {
+        // Hotmail also sets cookies for /owa/0/, /mail/0/, /calendar/0/ etc.,
+        // but we can use only the /owa/0/ cookie.
+        // We also need to use that URL for the service request.
+        // This needs to happen before CheckLoginFinished().
+          urlObj.hostname = cookie.domain;
+          urlObj.pathname = cookie.path;
+        }
+        if (!removed &&
```

Comment:
If `removed` actually happens, then you should check that at the top of the function and make an early exit. See above.